### PR TITLE
Narrow arrays in union based on count() with smaller/greater operator

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11480

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -244,12 +244,51 @@ public function specifyTypesInCondition(
...
+								continue;
+							}
+						}
+						if ($context->falsey()) {
+							if (!$isSize->yes()) {
+								continue;
+							}
+						}
+
+						$result[] = $innerType;
```
Comment (ondrejmirtes) [inline src/Analyser/TypeSpecifier.php]: I'd like some DRY around this, we already have the same code elsewhere. We could have a method that just accepts the narrowed array size as Type, and return SpecifiedTypes.

File: src/Analyser/TypeSpecifier.php [hunk 2]
```diff
@@ -244,12 +244,51 @@ public function specifyTypesInCondition(
...
+
+				if (
+					$isNormalCount->yes()
+					&& $argType instanceof UnionType
+					&& $leftType instanceof ConstantIntegerType
+				) {
+					if ($orEqual) {
+						$constantType = IntegerRangeType::createAllGreaterThanOrEqualTo($leftType->getValue());
+					} else {
+						$constantType = IntegerRangeType::createAllGreaterThan($leftType->getValue());
```
Comment (ondrejmirtes) [inline src/Analyser/TypeSpecifier.php]: We already have this elsewhere:

```php
					$context->true() && (IntegerRangeType::createAllGreaterThanOrEqualTo(1 - $offset)->isSuperTypeOf($leftType)->yes())
					|| ($context->false() && (new ConstantIntegerType(1 - $offset))->isSuperTypeOf($leftType)->yes())
```

Would be nice to reuse that.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/TypeSpecifier.php (hunk 2 above)
Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: I don't think its the same thing. in the first block we work on the concrete constant integer, used in the comparison while in the second block we just check for the comparison value beeing >0 or >1.

File: src/Analyser/TypeSpecifier.php (hunk 1 above)
Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: done

**Comment by ondrejmirtes** [comment]
Awesome, thank you!

**Pull request merged**